### PR TITLE
feat: thread task pool support revoke

### PR DESCRIPTION
**Pull request by yywing**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

### CODE REVIEW

File: celery/utils/thread_pool.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+from concurrent.futures import ThreadPoolExecutor as _T
+
+from celery.exceptions import WorkerTerminate
+
+
+def _async_raise(tid, exctype):
+    """raises the exception, performs cleanup if needed"""
+    if not inspect.isclass(exctype):
+        raise TypeError("Only types can be raised (not instances)")
+    res = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_long(tid),
```
Comment (yywing) [inline celery/utils/thread_pool.py:47]: TODO: Changed in version 3.7: The type of the id parameter changed from long to unsigned long.

Comment (thedrow) [reply celery/utils/thread_pool.py:47]: Please add a condition here and a TODO comment

File: celery/utils/thread_pool.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        self.raise_exc(WorkerTerminate)
+
+
+class ThreadPoolExecutor(_T):
+    """
+    1. Don't catch WorkerTerminate exception, unless you know what you do
+    2. The exception will be raised only when execution returns to the python code.
+    """
+    def _adjust_thread_count(self):
+        with mock.patch("threading.Thread", new=Thread):
```
Comment (thedrow) [inline celery/utils/thread_pool.py:92]: I don't really like this in production code.

**Review by thedrow — CHANGES_REQUESTED** [review]

**Comment by matusvalo** [comment]
I am skeptical about this PR. This PR kills a thread via triggering exception in other thread via Dark Magic. This can work when python code is executed during revoking. But in case when the code is blocked in I/O or is executing some C code the **revoking simply won't work**. E.g. revoking a task that is blocked by executing SQL query (e.g. due Database lock) will fail.

I am not aware about any universal way how to kill/terminate a thread. Hence, supporting revoking thread task pool can cause that for some users confusion when it simply fails but only sometimes.

**Comment by auvipy** [comment]
> I am skeptical about this PR. This PR kills a thread via triggering exception in other thread via Dark Magic. This can work when python code is executed during revoking. But in case when the code is blocked in I/O or is executing some C code the **revoking simply won't work**. E.g. revoking a task that is blocked by executing SQL query (e.g. due Database lock) will fail.
> 
> I am not aware about any universal way how to kill/terminate a thread. Hence, supporting revoking thread task pool can cause that for some users confusion when it simply fails but only sometimes.

I think we should close this PR.

**Comment by matusvalo** [comment]
Closed due limitations of killing a thread.

**Pull request closed without merging**